### PR TITLE
Policy gradient example: Support multi-GPU training

### PR DESCRIPTION
**Pull request by ericl**
This is an implementation of the "multi-tower" parallel synchronous SGD model, similar to https://github.com/tensorflow/tensorflow/blob/r0.7/tensorflow/models/image/cifar10/cifar10_multi_gpu_train.py

At the start of each iteration, we split the data across the devices and load it into device memory in a tensorflow variable. During each SGD minibatch gradients are computed over a slice of the loaded data on each device, averaged across devices, and the average applied to the shared model weights.

With large enough batch sizes this gives near-linear speedups on multiple GPUs. Performance is also greatly improved on a single GPU due to faster data loading.

Perf stats for the 1st iteration of Pong-v0, batch size = 4096
-----------------
1 GPU (master branch w extra shuffles removed)
sgd examples/s: 126.9107562781184

1 GPU (this branch)
rollouts time: 52.24811363220215
shuffle time: 2.329425811767578
load time: 3.2698168754577637
sgd time: 60.46086621284485
sgd examples/s: 706.6389001043344

2 GPUs
rollouts time: 53.51676917076111
shuffle time: 2.4285168647766113
load time: 3.6924355030059814
sgd time: 33.014596939086914
sgd examples/s: 1339.4075378714283

4 GPUs
rollouts time: 53.179341316223145
shuffle time: 2.3760111331939697
load time: 4.1257312297821045
sgd time: 23.697580337524414
sgd examples/s: 1838.4577403884966

8 GPUs
rollouts time: 55.861610651016235
shuffle time: 2.4734747409820557
load time: 5.357516765594482
sgd time: 16.425883293151855
sgd examples/s: 2693.4928983968507

16 GPUs
rollouts time: 53.64489507675171
shuffle time: 2.4809675216674805
load time: 7.8745152950286865
sgd time: 26.39092707633972
sgd examples/s: 1623.5503919986825

### CODE REVIEW

File: examples/policy_gradient/examples/example.py [hunk 1]
```diff
@@ -12,27 +12,34 @@
...
           "sgd_batchsize": 128,
           "entropy_coeff": 0.0,
           "clip_param": 0.3,
           "kl_target": 0.01,
-          "timesteps_per_batch": 40000,
+          "timesteps_per_batch": 4000,
           "num_agents": 5,
           "tensorboard_log_dir": "/tmp/ray",
-          "trace_level": tf.RunOptions.NO_TRACE}
+          "trace_level": tf.RunOptions.FULL_TRACE}
```
Comment (ericl) [inline examples/policy_gradient/examples/example.py]: Ignore all these test changes for now.

File: examples/policy_gradient/reinforce/models/fcnet.py [hunk 2]
```diff
@@ -17,19 +17,19 @@ def _initializer(shape, dtype=None, partition_info=None):
 
 
 def fc_net(inputs, num_classes=10, logstd=False):
-  with tf.name_scope("fc_net") as net:
+  with tf.name_scope("fc_net"):
```
Comment (ericl) [inline examples/policy_gradient/reinforce/models/fcnet.py]: The name scoping was messed up before, it should be fixed now.

File: examples/policy_gradient/reinforce/agent.py [hunk 3]
```diff
@@ -2,48 +2,185 @@
...
+      distribution_class = Categorical
+    else:
+      raise NotImplemented("action space" + str(type(action_space)) +
+                           "currently not supported")
+    self.prev_logits = tf.placeholder(tf.float32, shape=(None, self.logit_dim))
+
+    # Defines operations for staging input training data.
+    self.num_splits = len(devices)
+    with tf.device("/cpu:0"):
+      stage = StagingArea(
```
Comment (ericl) [inline examples/policy_gradient/reinforce/agent.py]: According to the docs, the staging area memory is located on the device on which it is created, and data copied to the destination device on area.get(). I wonder if it would be more efficient to create a staging area per device so that the copy happens on put() instead.

File: examples/policy_gradient/reinforce/agent.py [hunk 4]
```diff
@@ -2,48 +2,185 @@
...
+def average_gradients(tower_grads):
+  """Calculate the average gradient for each shared variable across all towers.
+  Note that this function provides a synchronization point across all towers.
+  Args:
+    tower_grads: List of lists of (gradient, variable) tuples. The outer list
+      is over individual gradients. The inner list is over the gradient
+      calculation for each tower.
+  Returns:
+     List of pairs of (gradient, variable) where the gradient has been averaged
+     across all towers.
```
Comment (pcmoritz) [inline examples/policy_gradient/reinforce/agent.py]: consider mentioning as a TODO here that we can potentially use NCCL (see https://www.tensorflow.org/performance/performance_models) in the future to speed things up here if it becomes a bottleneck.

File: examples/policy_gradient/reinforce/agent.py [hunk 5]
```diff
@@ -2,48 +2,185 @@
...
+        self.ppo_towers[0].loss,  # all towers have equivalent vars
+        self.sess)
+    self.observation_filter = MeanStdFilter(preprocessor.shape, clip=None)
+    self.reward_filter = MeanStdFilter((), clip=5.0)
     self.sess.run(tf.global_variables_initializer())
 
+  def stage_trajectory_data(self, batch):
+    self.sess.run(
+        self.stage_ops,
+        feed_dict={self.observations: make_divisible_by(batch["observations"], self.num_splits),
```
Comment (pcmoritz) [inline examples/policy_gradient/reinforce/agent.py]: this is potentially a little dangerous I think because it might drop data in every batch; I'd consider asserting that the batch size is a multiple of num_splits and either giving an error if it is not, or making it divisible and giving a warning

**Comment by robertnishihara** [comment]
Ok to test.

File: examples/policy_gradient/reinforce/agent.py [hunk 6]
```diff
@@ -2,48 +2,258 @@
...
+from reinforce.distributions import Categorical, DiagGaussian
 from reinforce.env import BatchedEnv
 from reinforce.policy import ProximalPolicyLoss
 from reinforce.filter import MeanStdFilter
 from reinforce.rollout import rollouts, add_advantage_values
+from reinforce.utils import make_divisible_by, average_gradients
+
+
+# Each tower is a copy of the policy graph pinned to a specific device
+Tower = namedtuple('Tower', ['init_op', 'grads', 'policy'])
```
Comment (robertnishihara) [inline examples/policy_gradient/reinforce/agent.py]: `'` -> `"`

File: examples/policy_gradient/reinforce/utils.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 from __future__ import division
 from __future__ import print_function
 
+import tensorflow as tf
```
Comment (robertnishihara) [inline examples/policy_gradient/reinforce/utils.py]: imports in alphabetical order

File: examples/policy_gradient/reinforce/utils.py [hunk 8]
```diff
@@ -32,17 +33,51 @@ def concatenate(weights_list):
...
-      batch[key] = trajectory[key][curr_index:(curr_index + batchsize)]
-    curr_index += batchsize
-    yield batch
+def make_divisible_by(array, n):
+  return array[0:array.shape[0] - array.shape[0] % n]
+
+
+def average_gradients(tower_grads):
+  """Calculate the average gradient for each shared variable across all towers.
+  Note that this function provides a synchronization point across all towers.
```
Comment (robertnishihara) [inline examples/policy_gradient/reinforce/utils.py]: - the first line of documentation should fit in one line
- add a newline before Args
- add a newline before Returns

File: examples/policy_gradient/reinforce/utils.py [hunk 9]
```diff
@@ -32,17 +33,51 @@ def concatenate(weights_list):
...
+  Args:
+    tower_grads: List of lists of (gradient, variable) tuples. The outer list
+      is over individual gradients. The inner list is over the gradient
+      calculation for each tower.
+  Returns:
+     List of pairs of (gradient, variable) where the gradient has been averaged
+     across all towers.
+
+  TODO(ekl) we could use NCCL if this becomes a bottleneck
```
Comment (robertnishihara) [inline examples/policy_gradient/reinforce/utils.py]: colon after TODO(ekl)

File: examples/policy_gradient/reinforce/agent.py [hunk 10]
```diff
@@ -2,48 +2,258 @@
...
+            self.actions: make_divisible_by(
+                trajectories["actions"].squeeze(), self.batch_size),
+            self.prev_logits: make_divisible_by(
+                trajectories["logprobs"], self.batch_size),
+        },
+        options=run_options,
+        run_metadata=run_metadata)
+    if full_trace:
+      trace = timeline.Timeline(step_stats=run_metadata.step_stats)
+      trace_file = open('/tmp/ray/timeline-load.json', 'w')
```
Comment (robertnishihara) [inline examples/policy_gradient/reinforce/agent.py]: `'` -> `"`

File: examples/policy_gradient/reinforce/agent.py [hunk 11]
```diff
@@ -2,48 +2,258 @@
...
+
+    tuples_per_device = len(truncated_obs) / len(self.devices)
+    assert tuples_per_device % self.per_device_batch_size == 0
+    return tuples_per_device
+
+  def run_sgd_minibatch(self, batch_index, kl_coeff, full_trace, file_writer):
+    """
+    Runs a SGD step over the batch with index batch_index as created by
+    load_rollouts_data(), updating local weights.
+
```
Comment (robertnishihara) [inline examples/policy_gradient/reinforce/agent.py]: - first line of documentation should fit in one line (and come right after """)
- Colon after returns and then the next stuff should be on the next line

File: examples/policy_gradient/reinforce/agent.py [hunk 12]
```diff
@@ -2,48 +2,258 @@
...
+        [self.train_op, self.mean_loss, self.mean_kl, self.mean_entropy],
+        feed_dict={
+            self.batch_index: batch_index,
+            self.kl_coeff: kl_coeff},
+        options=run_options,
+        run_metadata=run_metadata)
+
+    if full_trace:
+      trace = timeline.Timeline(step_stats=run_metadata.step_stats)
+      trace_file = open('/tmp/ray/timeline-sgd.json', 'w')
```
Comment (robertnishihara) [inline examples/policy_gradient/reinforce/agent.py]: `'` -> `"`

File: examples/policy_gradient/reinforce/agent.py [hunk 13]
```diff
@@ -2,48 +2,258 @@
...
+
+
+# Each tower is a copy of the policy graph pinned to a specific device
+Tower = namedtuple('Tower', ['init_op', 'grads', 'policy'])
 
 
 class Agent(object):
-  def __init__(self, name, batchsize, preprocessor, config, use_gpu):
-    if not use_gpu:
+  """
```
Comment (robertnishihara) [inline examples/policy_gradient/reinforce/agent.py]: There should be a short one line summary immediately after """

File: examples/policy_gradient/examples/example.py [hunk 14]
```diff
@@ -47,32 +58,36 @@
...
-  agent = Agent(mdp_name, 1, preprocessor, config, True)
+  agent = Agent(mdp_name, 1, preprocessor, config, False)
 
   kl_coeff = config["kl_coeff"]
 
   file_writer = tf.summary.FileWriter(
       '{}/trpo_{}_{}'.format(
-          config["tensorboard_log_dir"], mdp_name, datetime.today()),
+          config["tensorboard_log_dir"], mdp_name,
+          str(datetime.today()).replace(' ', '_')),
```
Comment (robertnishihara) [inline examples/policy_gradient/examples/example.py]: these should all be `"`

File: examples/policy_gradient/reinforce/agent.py [hunk 15]
```diff
@@ -2,48 +2,258 @@
...
+                all_acts.initializer,
+                all_plog.initializer]),
+          grads,
+          policy)
+
+  def load_data(self, trajectories, full_trace):
+    """
+    Bulk loads the specified trajectories into device memory. The data is
+    split equally across all the devices.
+
```
Comment (robertnishihara) [inline examples/policy_gradient/reinforce/agent.py]: There should be a short one line summary after the """

Also, the return documentation should be formatted as

```
Returns:
  The number of tuples loaded per device.
```

**Comment by ericl** [comment]
Hmm, should the flake8 command be catching these nits?

**Comment by robertnishihara** [comment]
flake8 is pretty lenient. I may add some other linters as well (e.g., pylint) to catch more things. For now, there are a couple conventions we've been following outside of the linters.
